In [2]:
import pandas as pd
import numpy as np
from pathlib import Path

In [3]:
df = pd.read_csv(
    "../data/raw/vehicle_registrations.csv",
    sep=";",
    encoding="utf-8",
    low_memory=False 
)

df.head()

,PERSON,REG_ADDR_KOATUU,OPER_CODE,OPER_NAME,D_REG,DEP_CODE,DEP,BRAND,MODEL,VIN,MAKE_YEAR,COLOR,KIND,BODY,PURPOSE,FUEL,CAPACITY,OWN_WEIGHT,TOTAL_WEIGHT,N_REG_NEW
0,P,5610100000,502,СКАСУВАННЯ ПЕРЕРЕЄСТРАЦІЇ (ПЕРЕРЕЄСТРАЦІЯ НА П...,01.01.26,12235,ТСЦ 0741,RENAULT,SCENIC,VF1JM0UD631760738,2004,СІРИЙ,ЛЕГКОВИЙ,УНІВЕРСАЛ,ЗАГАЛЬНИЙ,БЕНЗИН АБО ГАЗ,1998.0,1627,2180,ВК7858ІС
1,P,6325456200,315,ПЕРЕРЕЄСТРАЦІЯ ТЗ НА НОВ. ВЛАСН. ПО ДОГОВОРУ У...,01.01.26,12361,ТСЦ 6347,OPEL,ZAFIRA,W0L0TGF75X2256914,1999,СИНІЙ,ЛЕГКОВИЙ,УНІВЕРСАЛ,ЗАГАЛЬНИЙ,БЕНЗИН,1796.0,1395,1960,АХ9456НК
2,P,1810100000,100,ПЕРВИННА РЕЄСТРАЦIЯ Б/В ТЗ ПРИДБАНОГО В ТОРГОВ...,01.01.26,13960,ТСЦ 8048,NISSAN,LEAF,SJNFAAZE0U6023253,2014,БІЛИЙ,ЛЕГКОВИЙ,ХЕТЧБЕК,ЗАГАЛЬНИЙ,ЕЛЕКТРО,NaN,1505,1945,АА4364YЕ
3,P,2120481601,440,ПЕРЕРЕЄСТРАЦIЯ ПРИ ВТРАТІ СВIДОЦТВА ПРО РЕЄСТР...,01.01.26,12262,ТСЦ 2142,MERCEDES-BENZ,ML 350,4JGBB86E06A019229,2005,ЧОРНИЙ,ЛЕГКОВИЙ,УНІВЕРСАЛ,ЗАГАЛЬНИЙ,БЕНЗИН,3498.0,2135,2830,АО1090НІ
4,P,2123681301,105,ПЕРВИННА РЕЄСТРАЦІЯ НОВОГО ТЗ ПРИДБАНОГО В ТОР...,01.01.26,12292,ТСЦ 8043,VDV,CRF 250,LRSYCNL3XR0D53096,2024,БІЛИЙ,МОТОЦИКЛ,МОТОЦИКЛ,ЗАГАЛЬНИЙ,БЕНЗИН,250.0,159,330,АА3804JВ


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 315679 entries, 0 to 315678
Data columns (total 20 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   PERSON           315679 non-null  object 
 1   REG_ADDR_KOATUU  315679 non-null  int64  
 2   OPER_CODE        315679 non-null  int64  
 3   OPER_NAME        315679 non-null  object 
 4   D_REG            315679 non-null  object 
 5   DEP_CODE         315679 non-null  int64  
 6   DEP              315679 non-null  object 
 7   BRAND            315679 non-null  object 
 8   MODEL            315679 non-null  object 
 9   VIN              315646 non-null  object 
 10  MAKE_YEAR        315679 non-null  int64  
 11  COLOR            315679 non-null  object 
 12  KIND             315679 non-null  object 
 13  BODY             315679 non-null  object 
 14  PURPOSE          315679 non-null  object 
 15  FUEL             304920 non-null  object 
 16  CAPACITY         289646 non-null  floa

In [5]:
df.shape

(315679, 20)

In [6]:
df.isnull().sum().sort_values(ascending=False)

CAPACITY           26033
FUEL               10759
N_REG_NEW           4119
OWN_WEIGHT            44
VIN                   33
TOTAL_WEIGHT          16
COLOR                  0
PURPOSE                0
BODY                   0
KIND                   0
PERSON                 0
REG_ADDR_KOATUU        0
MODEL                  0
BRAND                  0
DEP                    0
DEP_CODE               0
D_REG                  0
OPER_NAME              0
OPER_CODE              0
MAKE_YEAR              0
dtype: int64

In [7]:
duplicates = df.duplicated().sum()
duplicates

710

In [8]:
df.describe()

,REG_ADDR_KOATUU,OPER_CODE,DEP_CODE,MAKE_YEAR,CAPACITY
count,3.156790e+05,315679.000000,315679.000000,315679.000000,289646.000000
mean,4.489821e+09,282.081105,12110.217537,2011.744519,2402.333051
std,2.414632e+09,103.163439,1007.005959,8.419481,2047.667406
min,1.101000e+08,18.000000,10000.000000,1900.000000,0.000000
25%,2.310100e+09,254.000000,12256.000000,2007.000000,1595.000000
50%,4.810100e+09,314.000000,12286.000000,2012.000000,1984.000000
75%,6.324285e+09,315.000000,12339.000000,2018.000000,2461.000000
max,8.536900e+09,570.000000,14930.000000,2026.000000,25860.000000


In [9]:
invalid_years = df[(df['MAKE_YEAR'] < 1900) | (df['MAKE_YEAR'] > 2026)]
print(f"Кількість записів з некоректним роком: {len(invalid_years)}")

zero_weight = df[df['OWN_WEIGHT'] == 0]
print(f"Кількість авто з нульовою вагою: {len(zero_weight)}")

Кількість записів з некоректним роком: 0
Кількість авто з нульовою вагою: 0


In [10]:
print("Унікальні значення палива:")
print(df['FUEL'].unique())

print(f"Унікальних марок (до очистки): {df['BRAND'].nunique()}")
print(f"Унікальних марок (після strip/upper): {df['BRAND'].str.strip().str.upper().nunique()}")

Унікальні значення палива:
['БЕНЗИН АБО ГАЗ' 'БЕНЗИН' 'ЕЛЕКТРО' 'ДИЗЕЛЬНЕ ПАЛИВО'
 'ЕЛЕКТРО АБО БЕНЗИН' 'ГАЗ' nan 'ЕЛЕКТРО АБО ДИЗЕЛЬНЕ ПАЛИВО'
 'БЕНЗИН, ГАЗ АБО ЕЛЕКТРО' 'ВІДСУТНЄ' 'НЕ ВИЗНАЧЕНО' '.'
 'ДИЗЕЛЬНЕ ПАЛИВО АБО ГАЗ']
Унікальних марок (до очистки): 1047
Унікальних марок (після strip/upper): 1047


## Підсумки перевірки якості даних (Data Quality Report)

Після проведення первинного аналізу набору даних про реєстрацію ТЗ було виявлено наступне:

### 1. Цілісність даних та пропуски
* **Загальна кількість записів:** 315,679 рядків.
* **Найбільша кількість пропусків** спостерігається у колонці `CAPACITY` (26,033) та `FUEL` (10,759). Пропуски в об'ємі двигуна часто корелюють з електромобілями, що є логічним для даного типу ТЗ.
* **Критичні дані:** Поля `BRAND`, `MODEL` та `MAKE_YEAR` не мають пропусків, що дозволяє проводити подальший аналіз популярності марок та віку авто.

### 2. Дублікати
* Виявлено **710 повних дублікатів**. Це свідчить про наявність технічних повторів у вивантаженні, які необхідно видалити перед етапом дослідження (Research).

### 3. Логічні аномалії та валідація
* **MAKE_YEAR:** Аномальних значень року (менше 1900 або більше 2026) не виявлено. Мінімальний рік у реєстрі — 1900, максимальний — 2026.
* **FUEL:** Виявлено 12 унікальних категорій палива. Присутні неінформативні значення, такі як `.` або `ВІДСУТНЄ`, які потребують додаткової обробки.
* **BRAND:** У реєстрі представлено 1047 унікальних марок. Перевірка через переведення в верхній регістр та видалення пробілів показала, що назви брендів уже уніфіковані (кількість залишилася незмінною — 1047).

### 4. Подальші кроки (Data Cleaning)
* Видалити виявлені дублікати.
* Привести типи даних (наприклад, масу авто) до числових форматів для розрахунків.
* Обробити пропуски в колонці `FUEL` та `CAPACITY` перед початком візуалізації.